### PR TITLE
adds name, value, SlackConfirmationField to slack action

### LINKED ISSUE #1556: [slack] message attachment actions implementation incomplete
**What did you do?**

Tried to attach action buttons to slack message as per https://api.slack.com/docs/message-buttons.
Slack considers the following attachments[].actions configuration valid, however the alertmanager doesn't:
```
{
...
"attachments": [
  ...
  "actions": [
    {
      "name": "<name>",
      "text": "<text>",
      "type": "<type>",
      "value": "<value"
    }
    ...
  ]
]
}
```

The current implementation of the [SlackAction](https://github.com/prometheus/alertmanager/blob/master/config/notifiers.go#L236-L241) allows only buttons with links.
I propose to extend this with the `name`, `value` attributes and make configuring these too as well as the `url` parameter optional as both configuration are valid actions as per above cited slack documentation.

**What did you expect to see?**

Buttons attached to slack message

**What did you see instead? Under which circumstances?**

Alertmanager refused to load configuration as the [SlackAction.URL parameter] (https://github.com/prometheus/alertmanager/blob/master/config/notifiers.go#L239) is missing: `FAILED: missing value in Slack url configuration`

**Environment**
* Alertmanager version: v0.15.2

### PR DESCRIPTION
**Pull request by auhlig**
Fixes #1556. Either `URL` or `Name` and `Value` need to be provided in slack action. LGTY @simonpasquier? Thanks in advance

### CODE REVIEW

**Comment by auhlig** [comment]
test failures seem unrelated. fails for current master on my machine - without above changes - as well

**Comment by simonpasquier** [comment]
Hmm, I'm definitely baffled by the lack of consistency in the Stack documentation. https://api.slack.com/docs/interactive-message-field-guide#action_fields says that `name`, `text` and `type` are mandatory while https://api.slack.com/docs/message-attachments#action_fields doesn't mention `name` and states that `url` is mandatory. I'd be reluctant to merge a change without more insights on the right way.

@stuartnelson3 any thoughts on this?

**Comment by auhlig** [comment]
I agree the slack API is quite confusing.
I understood, there are (at least) 2 types of buttons in slack: *message* and *link* buttons in a slack action.
The API says `type`, `text`, `url` are `..required when using link buttons` (https://api.slack.com/docs/message-attachments#action_fields). 
While *message* buttons as specified [here](https://api.slack.com/docs/interactive-message-field-guide#action_fields) need to have `name`, `text`, `type`.

File: config/notifiers.go [hunk 1]
```diff
@@ -252,9 +255,13 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if c.Text == "" {
 		return fmt.Errorf("missing value in Slack text configuration")
 	}
-	if c.URL == "" {
+	// either URL or Name and Value need to be provided
+	if c.URL != "" && c.Name != "" && c.Value != "" {
```
[the remark this answers is not in the document]
Comment (auhlig) [reply config/notifiers.go]: right. adjusted

File: config/notifiers.go [hunk 2]
```diff
@@ -252,9 +255,13 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if c.Text == "" {
 		return fmt.Errorf("missing value in Slack text configuration")
```
[the remark this answers is not in the document]
Comment (auhlig) [reply config/notifiers.go]: done

File: config/notifiers.go [hunk 3]
```diff
@@ -232,12 +232,16 @@ func (c *PagerdutyConfig) UnmarshalYAML(unmarshal func(interface{}) error) error
 
 // SlackAction configures a single Slack action that is sent with each notification.
 // Each action must contain a type, text, and url.
-// See https://api.slack.com/docs/message-attachments#action_fields for more information.
+// See https://api.slack.com/docs/message-attachments#action_fields and https://api.slack.com/docs/message-buttons
+// for more information
```
Comment (simonpasquier) [inline config/notifiers.go]: (nit) missing dot at the end

File: config/notifiers.go [hunk 4]
```diff
@@ -250,11 +254,47 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		return fmt.Errorf("missing type in Slack action configuration")
 	}
 	if c.Text == "" {
-		return fmt.Errorf("missing value in Slack text configuration")
+		return fmt.Errorf("missing text value in Slack text configuration")
 	}
-	if c.URL == "" {
-		return fmt.Errorf("missing value in Slack url configuration")
+	// either URL or Name and Value need to be provided
```
Comment (simonpasquier) [inline config/notifiers.go]: missing dot

File: config/notifiers.go [hunk 5]
```diff
@@ -250,11 +254,47 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+	if c.URL == "" && c.Name == "" {
+		return fmt.Errorf("missing name or url in Slack action configuration")
 	}
+	if (c.Name == "" || c.Value == "") && c.URL == "" {
```
Comment (simonpasquier) [inline config/notifiers.go]: I'm still not sure that it is the correct approach: AFAIU `value` isn't mandatory. Also what happens if the configuration defines `url` and `name` at the same time? What is the result in Slack?

Maybe do something like this instead?

```go
if c.URL != "" {
  // Clear all message action fields.
  c.Name = ""
  c.Value = ""
  c.ConfirmField = nil
} else if c.Name != "" {
  c.URL = ""
} else {
  return fmt.Errorf("missing name or url in Slack action configuration")
}
```

File: config/notifiers.go [hunk 6]
```diff
@@ -250,11 +254,47 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+// by asking them to confirm their button click one more time.
+// See https://api.slack.com/docs/interactive-message-field-guide#confirmation_fields for more information.
+type SlackConfirmationField struct {
+	Text        string `yaml:"text,omitempty"  json:"text,omitempty"`
+	Title       string `yaml:"title,omitempty"  json:"title,omitempty"`
+	OkText      string `yaml:"ok_text,omitempty"  json:"ok_text,omitempty"`
+	DismissText string `yaml:"dismiss_text,omitempty"  json:"dismiss_text,omitempty"`
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface for SlackConfirmationField
```
Comment (simonpasquier) [inline config/notifiers.go]: missing dot

File: config/notifiers.go [hunk 7]
```diff
@@ -250,11 +254,47 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+func (c *SlackConfirmationField) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type plain SlackConfirmationField
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+	if c.Text == "" {
```
Comment (simonpasquier) [inline config/notifiers.go]: Only `text` is required, the other fields are optional: https://api.slack.com/docs/interactive-message-field-guide#confirmation_fields

Comment (auhlig) [reply config/notifiers.go]: done

File: config/notifiers.go (hunk 6 above)
Comment (auhlig) [reply config/notifiers.go]: dotted

File: config/notifiers.go (hunk 4 above)
Comment (auhlig) [reply config/notifiers.go]: added dot

File: config/notifiers.go (hunk 3 above)
Comment (auhlig) [reply config/notifiers.go]: added dot

**Comment by auhlig** [comment]
Also extended tests. LGTY @simonpasquier?

**Comment by auhlig** [comment]
Anything else that needs to be done @simonpasquier?

File: config/notifiers.go [hunk 8]
```diff
@@ -250,10 +254,39 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		return fmt.Errorf("missing type in Slack action configuration")
 	}
 	if c.Text == "" {
-		return fmt.Errorf("missing value in Slack text configuration")
+		return fmt.Errorf("missing text value in Slack text configuration")
```
Comment (simonpasquier) [inline config/notifiers.go]: `"missing text in Slack action configuration"`

File: config/notifiers.go [hunk 9]
```diff
@@ -250,10 +254,39 @@ func (c *SlackAction) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface for SlackConfirmationField.
+func (c *SlackConfirmationField) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type plain SlackConfirmationField
+	if err := unmarshal((*plain)(c)); err != nil {
+		return err
+	}
+	if c.Text == "" {
+		return fmt.Errorf("missing text in slack action confirm")
```
Comment (simonpasquier) [inline config/notifiers.go]: `"missing text in slack configuration configuration"`

File: notify/impl.go [hunk 10]
```diff
@@ -747,6 +747,8 @@ func (n *Slack) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
 				Text:  tmplText(action.Text),
 				URL:   tmplText(action.URL),
 				Style: tmplText(action.Style),
+				Name:  tmplText(action.Name),
+				Value: tmplText(action.Value),
```
Comment (simonpasquier) [inline notify/impl.go]: you're missing the `ConfirmField` configuration.

**Comment by auhlig** [comment]
uses `tmplText()` on `SlackConfirmationField`. LGTY @simonpasquier?

**Comment by simonpasquier** [comment]
Thanks for your patience @auhlig. I'll give it a try as I happen to have access to a Slack instance...
Once it is merged, the documentation would need updates too.

File: notify/impl.go [hunk 11]
```diff
@@ -747,6 +747,14 @@ func (n *Slack) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
...
+				ConfirmField: &config.SlackConfirmationField{
```
Comment (simonpasquier) [inline notify/impl.go]: You need to test if `action.ConfirmField` is nil otherwise it will panic.

Comment (auhlig) [reply notify/impl.go]: done

**Comment by auhlig** [comment]
Added check and opened PR for docs @simonpasquier

**Comment by auhlig** [comment]
Not sure why (only) travis complains. Do I need to run `make assets` here?

**Comment by auhlig** [comment]
Tested with slack. Works for me (TM).
Will update https://github.com/prometheus/docs/pull/1196 today.
Would you mind taking another look @simonpasquier?

**Comment by auhlig** [comment]
Tested with slack. Works for me (TM).
Will update https://github.com/prometheus/docs/pull/1196 today.
Would you mind taking another look and consider merging @simonpasquier? :heart:

**Comment by auhlig** [comment]
Tested with slack. Works for me (TM).
Will update https://github.com/prometheus/docs/pull/1196 today.
Would you mind taking another look and consider merging @simonpasquier? :heart:

**Comment by auhlig** [comment]
CI is green. Tested with slack. Works for me (TM)
Will update the docs PR today.
Would you mind taking another look and can we consider merging @simonpasquier? :heart:

**Comment by simonpasquier** [comment]
> Did slack attachments ever work?

They surely did. TBH Slack is terrible at dealing with backward compatibility: things that used to work at time T may break at time T+N for some users and continue to work for the others. I wouldn't include https://github.com/prometheus/alertmanager/pull/1557/commits/f61c50dfe2a79eb164135ca418eae0083188e415 in this PR as it is unrelated. If it is broken with the current `master`, please file an issue with all relevant details and it should be addressed separately.

**Comment by auhlig** [comment]
Ewww slack -.-
Extracted the `callback_id` to https://github.com/prometheus/alertmanager/pull/1592 and reset this PR to the last approved commit. Sorry for the back- and forth. Is this good to go @simonpasquier?

File: config/notifiers.go [hunk 12]
```diff
@@ -247,12 +247,16 @@ func (c *PagerdutyConfig) UnmarshalYAML(unmarshal func(interface{}) error) error
 
 // SlackAction configures a single Slack action that is sent with each notification.
 // Each action must contain a type, text, and url.
```
[the remark this answers is not in the document]
Comment (auhlig) [reply config/notifiers.go]: it's gone

**Comment by simonpasquier** [comment]
:+1: cc @mxinden @stuartnelson3

**Comment by simonpasquier** [comment]
Yes thanks for the work @auhlig. That was more involved than expected!

**Comment by auhlig** [comment]
Thanks guys ❤️

**Pull request merged**